### PR TITLE
Fix #3491 and #3479: App layer mechanism to resume exploration 

### LINKED ISSUE #3479: Enable checkpointing once mechanism to resume exploration with checkpoints is implemented.
Checkpointing is disabled until the mechanism to resume explorations with checkpoints is implemented. 
Once resuming mechanism is complete checkpointing will have to be enabled by changing the value of the boolean shouldSavePartialProgress in the following classes:
1. TopicLessonsFragment.kt
2. RecentlyPlayedFragmentPresenter.kt
3. StoryChapterSummaryViewModel.kt

### LINKED ISSUE #3491: Modify app layer classes to start exploration
The app layer classes which start playing an exploration have to be modified to start the exploration with saved progress.

Target | Date
--|--
**PR creation:** | 2021-08-07
**PR completion:** | 2021-08-11

### PR DESCRIPTION
**Pull request by MaskedCarrot**
## Explanation

Fixes #3491 and #3479 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

#### StateFragmentTest
Screenshot from 2021-08-21 04-19-15

#### ExplorationActivityTest
Screenshot from 2021-08-21 04-47-49

#### TopicLessonsFragmentTest
Screenshot from 2021-08-21 18-37-18

### CODE REVIEW

**Comment by MaskedCarrot** [comment]
closing this PR because it contains a lot of unnecessary changes, the new PR is #3672

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt [hunk 1]
```diff
@@ -221,17 +226,96 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
...
+        internalProfileId,
+        promotedStory.topicId,
+        promotedStory.storyId,
+        promotedStory.explorationId,
+        shouldSavePartialProgress,
+        canExplorationBeResumed = false,
+        backflowScreen = null
+      )
+    }
   }
```
Comment (MaskedCarrot) [inline app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: There are two erros due to this code:

1. If the exploration version doesn't match, both resume lesson and exploration activity are opened one after the other, when only one should have been opened.


I think the reason because exploration Activity and ResumeLessonActivity is starting are starting (one after the other) if the exploration doesn't match is since when the exploration versions don't match we start first call play exploration (that loads exploration and then saved checkpoint) and then open exploration activity, in the mean time version of exploration is updated (because play exploration saves a new checkpoint), the observer is notified and this time resume exploration activity is opened.




2. This error will surely not occur in real-world scenarios so I am not sure that if we should take care of this: if the exploration version is updated while the user is in the exploration (cannot occur in real-world scenarios), when the user clicks back to go back to the topic lessons fragment, upon returning to the topic lessons fragment they are automatically taken back to the exploration.

I am not sure why this is happening, but I think it might be because the version is now updated so the observer (that matched the version in the topic lessons fragment) starts the exploration again.

**Review by MaskedCarrot — COMMENTED** [review]
@BenHenning PTAL

File: domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt [hunk 2]
```diff
@@ -228,6 +235,31 @@ class ExplorationCheckpointController @Inject constructor(
...
+
+        if (checkpoint != null) {
+          AsyncResult.success(
+            chapterDataJsonObject?.getInt("version") == checkpoint.explorationVersion
+          )
+        } else
+          AsyncResult.failed(
+            ExplorationCheckpointNotFoundException(
+              "Checkpoint with the explorationId $explorationId was not found " +
+                "for profileId ${profileId.internalId}."
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt]:
```suggestion
              "Checkpoint with the exploration $explorationId was not found " +
                "for profile ${profileId.internalId}."
```

File: domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt [hunk 3]
```diff
@@ -228,6 +235,31 @@ class ExplorationCheckpointController @Inject constructor(
     }
   }
 
+  fun isSavedCheckpointCompatibleWithExploration(
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt]: Needs KDoc & tests.

File: domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt [hunk 4]
```diff
@@ -228,6 +235,31 @@ class ExplorationCheckpointController @Inject constructor(
...
+    profileId: ProfileId,
+    explorationId: String
+  ): DataProvider<Boolean> {
+    return retrieveCacheStore(profileId)
+      .transformAsync(
+        CHECK_IS_EXPLORATION_CHECKPOINT_COMPATIBLE_WITH_EXPLORATION_DATA_PROVIDER_ID
+      ) { explorationCheckpointDatabase ->
+        val checkpoint = explorationCheckpointDatabase.explorationCheckpointMap[explorationId]
+        val chapterDataJsonObject = jsonAssetRetriever.loadJsonFromAsset("$explorationId.json")
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt]: We should be using ExplorationRetriever here, instead, so that this can work correctly when loading from proto (we currently support loading topic data from both protos and json).

File: app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt [hunk 5]
```diff
@@ -33,16 +38,100 @@ class StoryChapterSummaryViewModel(
...
+        backflowScreen
+      )
+    } else {
+      playExploration(
+        canExplorationBeResumed,
+        shouldSavePartialProgress,
+        backflowScreen
+      )
+    }
+  }
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt]: If we need to check at all locations whether there is compatibility, and in that case we just abandon the checkpoint & start over, why not move the version check into the DataProvider that provides the checkpoint, return a default, and short-circuit to ExplorationActivity if no checkpoint was found?

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 6]
```diff
@@ -641,6 +647,9 @@ class TopicListController @Inject constructor(
         storyRecord.chaptersList.firstOrNull()?.let {
           nextChapterName = it.title
           explorationId = it.explorationId
+          // ChapterPlayState will be NOT_STARTED because this function only recommends first story
+          // of un-started topics.
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]:
```suggestion
          // ChapterPlayState will be NOT_STARTED because this function only recommends the first
          // story of un-started topics.
```

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt (hunk 1 above)
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: My suggestion elsewhere to change the flow to navigate based on the state of the checkpoint DataProvider might solve this (presuming we're doing that in ExplorationFragment) since we wouldn't be passing in an actual checkpoint to initiate the exploration if it's incompatible. Merging the two DataProviders together should avoid the race condition described here.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @MaskedCarrot. Took an initial pass (primarily on my codeowner files).

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 6 above)
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: done

File: domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt (hunk 4 above)
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt]: done

File: domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt (hunk 2 above)
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt]: done

File: domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt (hunk 3 above)
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt]: done

File: app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt (hunk 5 above)
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt]: As per our chat, used `removeObserver`

**Comment by BenHenning** [comment]
Unless I'm missing something, isn't this still using 2 DataProviders @MaskedCarrot? Per our discussion, my understanding is that we should use just 1 to simplify the dataflow and reduce potential inconsistencies (though the one regarding the observer is unavoidable without using removeObserver as you mentioned).

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt (hunk 1 above)
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: as per chats in hangouts, used removeObserver.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 7]
```diff
@@ -641,6 +647,9 @@ class TopicListController @Inject constructor(
         storyRecord.chaptersList.firstOrNull()?.let {
           nextChapterName = it.title
           explorationId = it.explorationId
+          // ChapterPlayState will be NOT_STARTED because this function only recommends the first
+          // story of un-started topics.
+          chapterPlayState = ChapterPlayState.NOT_STARTED
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: Should be moved outside the ``let`` block since it's unrelated.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt [hunk 8]
```diff
@@ -690,7 +700,8 @@ class TopicListController @Inject constructor(
     totalChapterCount: Int,
     nextChapterName: String?,
     explorationId: String?,
-    isTopicConsideredCompleted: Boolean
+    isTopicConsideredCompleted: Boolean,
+    chapterProgress: ChapterProgress?
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: Can this name be more specific? Maybe ``nextChapterProgress``?

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt [hunk 9]
```diff
@@ -93,6 +96,15 @@ class ExplorationStorageModuleTest {
...
     fun provideCacheAssetsLocally(): Boolean = false
+
+    @Provides
+    @TopicListToCache
+    fun provideTopicListToCache(): List<String> = listOf()
+
+    @Provides
+    @LoadLessonProtosFromAssets
+    fun provideLoadLessonProtosFromAssets(testEnvironmentConfig: TestEnvironmentConfig): Boolean =
+      testEnvironmentConfig.isUsingBazel()
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt]: Out of curiosity, why are these needed now? I wouldn't expect these to be needed for a test suite that's just verifying the module providers.

File: testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt [hunk 10]
```diff
@@ -13,23 +13,37 @@ import org.mockito.MockitoAnnotations
...
 import org.oppia.android.util.data.DataProvider
 import org.oppia.android.util.data.DataProviders.Companion.toLiveData
 import javax.inject.Inject
 import javax.inject.Singleton
 
-const val FAKE_EXPLORATION_ID_1 = "fake_exploration_id_1"
-const val FAKE_EXPLORATION_TITLE_1 = "Fake exploration title_1"
-const val FAKE_EXPLORATION_ID_2 = "fake_exploration_id_2"
-const val FAKE_EXPLORATION_TITLE_2 = "Fake exploration title_2"
+const val FRACTIONS_EXPLORATION_0_TITLE = "What is a Fraction?"
```
Comment (BenHenning) [inline testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt:28]: Please add KDocs for all currently undocumented non-private members in this class & remove the exemption from scripts/assets/kdoc_validity_exemptions.textproto.

File: testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt [hunk 11]
```diff
@@ -49,36 +63,126 @@ class ExplorationCheckpointTestHelper @Inject constructor(
...
+      checkpoint
+    )
+    verifyProviderFinishesWithSuccess(saveCheckpointDataProvider)
+  }
+
+  /**
+   * Updates the saved checkpoint for Fractions, story 0, exploration 0. For this function to work
+   * correctly it should be called after [saveCheckpointForFractionsStory0Exploration0].
+   *
+   * @param profileId the profileID for which the checkpoint has to be saved
```
Comment (BenHenning) [inline testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt]:
```suggestion
   * @param profileId the profile ID for which the checkpoint has to be saved
```

Ditto elsewhere in the PR.

File: testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt [hunk 12]
```diff
@@ -79,20 +89,77 @@ class ExplorationCheckpointTestHelperTest {
...
+    testCoroutineDispatchers.runCurrent()
+
+    // Verify exploration checkpoint was saved with correct exploration title.
+    verify(mockExplorationCheckpointObserver, atLeastOnce())
+      .onChanged(explorationCheckpointCaptor.capture())
+    assertThat(explorationCheckpointCaptor.value.isSuccess()).isTrue()
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().explorationTitle)
+      .isEqualTo(FRACTIONS_EXPLORATION_0_TITLE)
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().pendingStateName)
+      .isEqualTo(FRACTIONS_STORY_0_EXPLORATION_0_SECOND_STATE_NAME)
```
Comment (BenHenning) [inline testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt]: Could you also add a verification line for state name in the previous test? Otherwise it seems a bit like update*() could do nothing & both tests would still pass (it's not obvious what the "difference" state being verified here is).

Ditto for the other similar tests here.

File: testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt [hunk 13]
```diff
@@ -79,20 +89,77 @@ class ExplorationCheckpointTestHelperTest {
...
     ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
   }
 
   @Test
-  fun testSaveFakeExplorationCheckpoint_checkCheckpointIsSaved() {
-    explorationCheckpointTestHelper.saveFakeExplorationCheckpoint(profileId.internalId)
+  fun testSaveCheckpointForFractionsStory0Exploration0_checkCheckpointIsSaved() {
+    explorationCheckpointTestHelper.saveCheckpointForFractionsStory0Exploration0(
+      profileId = profileId,
+      version = FRACTIONS_STORY_0_EXPLORATION_0_CORRECT_VERSION
```
Comment (BenHenning) [inline testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt]: Here & elsewhere: should we also be verifying that the version is used correctly (i.e. saved in the checkpoint)?

Suggest also adding a test case for a different version being used.

File: testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt [hunk 14]
```diff
@@ -102,43 +169,90 @@ class ExplorationCheckpointTestHelperTest {
...
+    testCoroutineDispatchers.runCurrent()
+
+    // Verify exploration checkpoint was saved with correct exploration title.
+    verify(mockExplorationCheckpointObserver, atLeastOnce())
+      .onChanged(explorationCheckpointCaptor.capture())
+    assertThat(explorationCheckpointCaptor.value.isSuccess()).isTrue()
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().explorationTitle)
+      .isEqualTo(FRACTIONS_EXPLORATION_0_TITLE)
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().pendingStateName)
+      .isEqualTo(FRACTIONS_STORY_0_EXPLORATION_0_SECOND_STATE_NAME)
```
Comment (BenHenning) [inline testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt]: This & above seem odd--why does a ratios checkpoint have fractions values in it?

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt [hunk 15]
```diff
@@ -252,17 +295,45 @@ class ExplorationCheckpointControllerTest {
...
+    )
+    explorationCheckpointController.retrieveExplorationCheckpoint(
+      firstTestProfile,
+      FRACTIONS_EXPLORATION_ID_0
+    ).toLiveData().observeForever(mockExplorationCheckpointObserver)
+    verifyMockObserverIsSuccessful(mockExplorationCheckpointObserver, explorationCheckpointCaptor)
+  }
+
+  @Test
+  fun testController_saveInCompatibleCheckpoint_checkpointCheckpointIsNotCompatible() {
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt]:
```suggestion
  fun testController_saveIncompatibleCheckpoint_checkpointCheckpointIsNotCompatible() {
```

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt [hunk 16]
```diff
@@ -252,17 +295,45 @@ class ExplorationCheckpointControllerTest {
...
+    explorationCheckpointTestHelper.saveCheckpointForFractionsStory0Exploration0(
+      profileId = firstTestProfile,
+      version = FRACTIONS_STORY_0_EXPLORATION_0_INCORRECT_VERSION
+    )
+    explorationCheckpointController.retrieveExplorationCheckpoint(
+      firstTestProfile,
+      FRACTIONS_EXPLORATION_ID_0
+    ).toLiveData().observeForever(mockExplorationCheckpointObserver)
+    verifyMockObserverIsFailure(mockExplorationCheckpointObserver, explorationCheckpointCaptor)
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt:357]: For this & all other tests: I suggest verifying the specific failure that's occurring to make sure that it's the failure you expect it to be.

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt (hunk 15 above)
Comment (BenHenning) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt]: Also, typo in name? 'checkpointCheckpoint'

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt [hunk 17]
```diff
@@ -252,17 +295,45 @@ class ExplorationCheckpointControllerTest {
...
+      FRACTIONS_EXPLORATION_ID_0
+    ).toLiveData().observeForever(mockExplorationCheckpointObserver)
+    verifyMockObserverIsSuccessful(mockExplorationCheckpointObserver, explorationCheckpointCaptor)
+  }
+
+  @Test
+  fun testController_saveInCompatibleCheckpoint_checkpointCheckpointIsNotCompatible() {
+    explorationCheckpointTestHelper.saveCheckpointForFractionsStory0Exploration0(
+      profileId = firstTestProfile,
+      version = FRACTIONS_STORY_0_EXPLORATION_0_INCORRECT_VERSION
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt]: Suggest instead: ``FRACTIONS_STORY_0_EXPLORATION_0_OLD_VERSION`` and ``FRACTIONS_STORY_0_EXPLORATION_0_CURRENT_VERSION`` names to be more specific.

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt [hunk 18]
```diff
@@ -221,17 +224,71 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
...
+      val explorationCheckpointLiveData =
+        explorationCheckpointController.retrieveExplorationCheckpoint(
+          ProfileId.getDefaultInstance(),
+          promotedStory.explorationId
+        ).toLiveData()
+
+      explorationCheckpointLiveData.observe(
+        fragment,
+        object : Observer<AsyncResult<ExplorationCheckpoint>> {
+          override fun onChanged(it: AsyncResult<ExplorationCheckpoint>?) {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Is the null property actually necessary? I don't think we do this in other observers.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @MaskedCarrot. Took a full review per my codeowners and left some comments--PTAL. Also, please fix the failing CI checks.

Overall, I think the mechanism looks good.

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt (hunk 9 above)
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt]: there are not needed here, removed them.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 8 above)
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: updated name.

File: domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt (hunk 7 above)
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt]: moved outside `let`

File: testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt (hunk 10 above)
Comment (MaskedCarrot) [reply testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt:28]: done

File: testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt (hunk 11 above)
Comment (MaskedCarrot) [reply testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt]: done

File: testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt (hunk 14 above)
Comment (MaskedCarrot) [reply testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt]: updated.

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt (hunk 15 above)
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt]: updated test name

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt (hunk 16 above)
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt:357]: done

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt (hunk 17 above)
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt]: done

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt (hunk 18 above)
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: removed the null property

File: testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt (hunk 12 above)
Comment (MaskedCarrot) [reply testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt]: done

File: testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt (hunk 13 above)
Comment (MaskedCarrot) [reply testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt]: we can only retrieve checkpoints if the version saved in the checkpoint matches with the current exploration version. Otherwise, a failure result is returned.

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt (hunk 9 above)
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt]: Added these back. Removing this is causing the tests to fail because we are injecting explorationcheckpointController here, which uses these dependencies.

**Review by vinitamurthi — APPROVED** [review]
LGTM on codeowner files!

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt [hunk 19]
```diff
@@ -221,17 +224,69 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
...
+          ProfileId.getDefaultInstance(),
+          promotedStory.explorationId
+        ).toLiveData()
+
+      explorationCheckpointLiveData.observe(
+        fragment,
+        object : Observer<AsyncResult<ExplorationCheckpoint>> {
+          override fun onChanged(it: AsyncResult<ExplorationCheckpoint>) {
+            if (it.isSuccess()) {
+              explorationCheckpointLiveData.removeObserver(this)
```
Comment (aggarwalpulkit596) [inline app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:245]: Shouldn't we call it outside this if and else condition instead?

File: app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt [hunk 20]
```diff
@@ -89,6 +90,9 @@ class TopicTestActivityForStoryTest {
   @Inject
   lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
 
+  @Inject
+  lateinit var explorationRetriever: ExplorationRetriever
```
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt]: Do we need this?

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt [hunk 21]
```diff
@@ -93,6 +95,11 @@ class ExplorationStorageModuleTest {
     @CacheAssetsLocally
     @Provides
     fun provideCacheAssetsLocally(): Boolean = false
+
+    @Provides
+    @LoadLessonProtosFromAssets
+    fun provideLoadLessonProtosFromAssets(testEnvironmentConfig: TestEnvironmentConfig): Boolean =
```
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt:101]: Are we using this?

**Review by anandwana001 — CHANGES_REQUESTED** [review]
nit thoughts.

Also, add a screenshot of espresso test result for all the test files from sharedTest directory.

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt (hunk 21 above)
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt:101]: yes, we inject explorationCheckpointController here and we use these there.

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt (hunk 19 above)
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:245]: I think moving this outside the if-else will make it trigger when the async result becomes pending

File: app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt (hunk 20 above)
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt]: remvoed

**Review by rt4914 — APPROVED** [review]
LGTM thanks.

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt (hunk 16 above)
Comment (BenHenning) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt:357]: Thanks--I think that this really improved the clarity of the tests!

File: domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt (hunk 21 above)
Comment (BenHenning) [reply domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt:101]: We also generally always want to use the switch logic when possible to ensure tests are verifying both the JSON & proto pipelines (in Gradle & Bazel, respectively).

File: testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt [hunk 22]
```diff
@@ -24,19 +24,44 @@ import org.oppia.android.util.data.DataProviders.Companion.toLiveData
...
+/** The exploration title of Ratios topic, story 0, exploration 1. */
 const val RATIOS_EXPLORATION_0_TITLE = "What is a Ratio?"
-const val FRACTIONS_STORY_0_EXPLORATION_0_CORRECT_VERSION = 85
-const val FRACTIONS_STORY_0_EXPLORATION_0_INCORRECT_VERSION = 25
-const val FRACTIONS_STORY_0_EXPLORATION_1_CORRECT_VERSION = 86
+
+/** The current exploration version of Fractions topic, story 0, exploration 0. */
+const val FRACTIONS_STORY_0_EXPLORATION_0_CURRENT_VERSION = 85
+
+/** The old exploration version of Fractions topic, story 0, exploration 0. */
```
Comment (BenHenning) [inline testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt]:
```suggestion
/** An old, not up-to-date version of Fractions topic, story 0, exploration 0. */
```

**Review by BenHenning — COMMENTED** [review]
Thanks @MaskedCarrot. Overall, PR LGTM other than a few nits--PTAL.

**Comment by anandwana001** [comment]
> nit thoughts.
> 
> Also, add a screenshot of espresso test result for all the test files from sharedTest directory.

Please add screenshot @MaskedCarrot

**Review by BenHenning — COMMENTED** [review]
Thanks @MaskedCarrot. LGTM for codeowners.

Please assign back once based on develop & with passing CI.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Test LGTM, but waiting for PR to get base branch update and test pass.

**Review by anandwana001 — APPROVED** [review]
All tests are passing on CI, LGTM for code owners' files.

**Review by BenHenning — APPROVED** [review]
Thanks @MaskedCarrot. LGTM as well.

**Comment by BenHenning** [comment]
Given all reviewers have approved & there are no outstanding conversations, merging.
